staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -62,18 +69,27 @@ var (
 		[]string{"verb", "group", "version", "resource", "subresource", "scope", "component"},
 	)
 	requestLatencies = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "apiserver_request_latency_seconds",
+			Help: "Response latency distribution in seconds for each verb, group, version, resource, subresource, scope and component.",
+			// Use buckets ranging from 125 ms to 8 seconds.
+			Buckets: prometheus.ExponentialBuckets(0.125, 2.0, 7),
+		},
+		[]string{"verb", "group", "version", "resource", "subresource", "scope", "component"},
+	)
+	deprecatedRequestLatencies = prometheus.NewHistogramVec(
 		prometheus.HistogramOpts{
 			Name: "apiserver_request_latencies",
-			Help: "Response latency distribution in microseconds for each verb, group, version, resource, subresource, scope and component.",
+			Help: "(Deprecated) Response latency distribution in microseconds for each verb, group, version, resource, subresource, scope and component.",
 			// Use buckets ranging from 125 ms to 8 seconds.
 			Buckets: prometheus.ExponentialBuckets(125000, 2.0, 7),
 		},
 		[]string{"verb", "group", "version", "resource", "subresource", "scope", "component"},
 	)
-	requestLatenciesSummary = prometheus.NewSummaryVec(
+	deprecatedRequestLatenciesSummary = prometheus.NewSummaryVec(
```

Comment:
@danielqsj - this metric was deprecated without adding anything new instead.
Was that intentional or just overlooked?

Reply:
The intention was that we felt the latency metrics as summaries are too expensive and not aggregatable across instances so just having the histograms should be sufficient. If there are other voices I'm happy to reconsider. As I mentioned above, I'm aware that scalability tests currently use this metric (or at least were), I recall us having had a discussion a year or so ago that the scalability tests should be moving away from using these (my apologies if I'm mixing up conversations, and this wasn't with you).

Reply:
The context is that this PR broke the tests, so we temporarily reverted to using the old metric: https://github.com/kubernetes/kubernetes/pull/73524

I think I roughly understand the concerns, but I would like to discuss what exactly we are trying to achieve and how we can do that without having a summary.
So the requirements are:
1. We need to be able to say whether 99th percentile was higher or lower than X (X being 1s, 5s or 30s depending on the calls - the exact SLO is described here: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md)
2. As there is defined in SLI, we would like to have the value from last 5 minutes.
3. I would like to be able to spot regressions (at the very least by looking post factum into graphs)
4. Aggregating across few masters is probably also an important thing.

@brancz - are those reasonable expectations?

IIUC, the first one we can achieve by appropriately choosing buckets (if we would have buckets that finish at 1s, 5s and 30s, we would be done).
The third one is probably also doable - we will just show on the graph % of requests handled for a given buckets - if that is growing, then it's probably a regression.
The question is: what about number (2). Do you have any recommendations?

Any other thoughts?

Reply:
As these are Prometheus metrics, assuming you're collecting these metrics with Prometheus, I see no problem at all for all of the points above:

1) exactly
2) we can see historic 99th latency with:

```
histogram_quantile(0.99, sum without(instance, pod) (rate(apiserver_request_latencies_bucket{job="apiserver"}[5m])))
```

> This assumes the prometheus config you are using has the `job="apiserver"`, `pod`, and `instance` labels. If you choose to have a different labelling, then those may vary, but the idea is the same.

3) the above calculates this over time
4) the above already does the latency aggregated over all instances (with today's summaries that's fundamentally not possible which is why we preferred the histogram :) )

The only thing that's different right now is that the buckets do not go up to 30s, they go up to 8s. If you have data to back up relevance up to 30s, that would be wonderful and I'd be more than happy to change the bucketing according to real world experience from the scalability tests! :)

Reply:
@brancz - thanks

I have more question about buckets. Can we customize buckets more?
In particular, there is huge difference for us between say 500ms and 900ms (while this is the same bucket).
What I would ideally like to see is probably sth like:
- every 100ms (or even a bit more) up to 1s
- then every 500ms up to 5s
- then maybe every 5s or so up to 30s

Regaring 30s - yes in large clusters having 150k pods, listing all pods may easily take more than 10s.

Reply:
Buckets are completely free form. Linear, log, or exponential buckets are just standards that are available to be used more easily. If we have a better idea, that's absolutely possible.

The godoc comment is literally:

```
	// Buckets defines the buckets into which observations are counted. Each
	// element in the slice is the upper inclusive bound of a bucket. The
	// values must be sorted in strictly increasing order. There is no need
	// to add a highest bucket with +Inf bound, it will be added
	// implicitly. The default value is DefBuckets.
	Buckets []float64
```

Reply:
Anyway - I opened https://github.com/kubernetes/kubernetes/pull/73638 for that.

Reply:
Looking at the cardinality we’re looking at creating roughly 300 histograms and with 50 buckets that’s 15k metrics. That’s a lot, but I feel this is important enough of a metric for it to be worth it.